### PR TITLE
Feature/lm 2250 lazy load functions

### PR DESCRIPTION
**Pull request by anton-locomote**
- Added fixes for @tttmaximttt PR comments.
- Changed indentation to spaces in lib/LazyLoad.js

### CODE REVIEW

File: lib/LazyLoad.js [hunk 1]
```diff
@@ -1,73 +1,131 @@
...
+    Object.defineProperty(Instance, "remove" + method + 'Async', {
+        value: function () {
+            return new Promise (function (resolve, reject) {
+                Model.find(conditions, {identityCache: false}).only(Model.id.concat(property)).first(function (err, item) {
+                    if (err) {
+                        return reject(err);
+                    }
+
+                    if (!item) {
+                        return reject(null);
```
Comment (tttmaximttt) [inline lib/LazyLoad.js]: Should be changed to resolve(null);In this case when we haven't any item this is not error is just bad request or request of not existing stuff, Promise shouldn't go in `catch`.

File: lib/LazyLoad.js [hunk 2]
```diff
@@ -1,73 +1,131 @@
...
+            });
+        },
+        enumerable: false
+    });
+
+    Object.defineProperty(Instance, "remove" + method + 'Async', {
+        value: function () {
+            return new Promise (function (resolve, reject) {
+                Model.find(conditions, {identityCache: false}).only(Model.id.concat(property)).first(function (err, item) {
+                    if (err) {
```
Comment (tttmaximttt) [inline lib/LazyLoad.js]: Use flatten `if` statement. https://github.com/airbnb/javascript/tree/es5-deprecated/es5#blocks

File: test/integration/property-lazyload.js [hunk 3]
```diff
@@ -82,51 +86,110 @@ describe("LazyLoad properties", function() {
...
+    it("promise-based getAccessor should return property", function (done) {
       Person.find().first(function (err, John) {
         should.equal(err, null);
+        John.should.be.a.Object();
+        John.getPhotoAsync()
+          .then(function (photo) {
+            photo.toString().should.equal(PersonPhoto.toString());
+            done();
+          }).catch(function(err) {
+            if (err) done(err);
```
Comment (tttmaximttt) [inline test/integration/property-lazyload.js]: you don't need check if err exist in this case, if Promise in have some err they always be in `catch`, so `catch` always has err. In you test case you when you have error you just need call `done` with err

File: test/integration/property-lazyload.js [hunk 4]
```diff
@@ -82,51 +86,110 @@ describe("LazyLoad properties", function() {
...
+              should.equal(err, null);
+              John.should.be.a.Object();
+
+              John.getPhotoAsync()
+                .then(function (photo) {
+                  should.equal(err, null);
+                  photo.toString().should.equal(OtherPersonPhoto.toString());
+                  done();
+                }).catch(function (err) {
+                  if (err) done(err);
```
Comment (tttmaximttt) [inline test/integration/property-lazyload.js]: The same about condition

File: test/integration/property-lazyload.js [hunk 5]
```diff
@@ -82,51 +86,110 @@ describe("LazyLoad properties", function() {
...
-
-              return done();
-            });
+            John.getPhotoAsync()
+              .then(function (photo) {
+                should.equal(err, null);
+                should.equal(photo, null);
+                done();
+              }).catch(function (err) {
+                if (err) done(err);
```
Comment (tttmaximttt) [inline test/integration/property-lazyload.js]: and here

File: lib/LazyLoad.js (hunk 2 above)
Comment (anton-locomote) [inline lib/LazyLoad.js]: agree && done

File: lib/LazyLoad.js (hunk 1 above)
Comment (anton-locomote) [inline lib/LazyLoad.js]: agree && done

File: test/integration/property-lazyload.js (hunk 3 above)
Comment (anton-locomote) [inline test/integration/property-lazyload.js]: agree && done

File: test/integration/property-lazyload.js (hunk 5 above)
Comment (anton-locomote) [inline test/integration/property-lazyload.js]: agree && done

File: test/integration/property-lazyload.js (hunk 4 above)
Comment (anton-locomote) [inline test/integration/property-lazyload.js]: agree && done

**Comment by dxg** [comment]
Same comment as in https://github.com/locomote/node-orm2/pull/1

File: lib/LazyLoad.js [hunk 6]
```diff
@@ -8,11 +11,27 @@ exports.extend = function (Instance, Model, properties) {
 
 function addLazyLoadProperty(name, Instance, Model, property) {
   var method = ucfirst(name);
+  var promiseFunctionPrefix = Settings.defaults().promiseFunctionPrefix;
```
Comment (dxg) [inline lib/LazyLoad.js]: I recommend accessing settings via `Model` as in https://github.com/locomote/node-orm2/blob/master/lib/Associations/Many.js#L129

The idea with settings is that they start with defaults, and can be overridden at [connection time](https://github.com/locomote/cbt/blob/master/app/models/_connection.coffee#L27) and even at [model definition](https://github.com/dresende/node-orm2/wiki/Defining-Models) time.

File: lib/LazyLoad.js [hunk 7]
```diff
@@ -8,11 +11,27 @@ exports.extend = function (Instance, Model, properties) {
...
+  var funcNamesObj = {
```
Comment (dxg) [inline lib/LazyLoad.js]: We don't need `Obj` in the name; we can just call this `functionNames`.

File: lib/LazyLoad.js [hunk 8]
```diff
@@ -8,11 +11,27 @@ exports.extend = function (Instance, Model, properties) {
...
+      callbackFuncName: "get" + method,
+      promiseFuncName: "get" + method + promiseFunctionPrefix
+    },
+    remove: {
+      callbackFuncName: "remove" + method,
+      promiseFuncName: "remove" + method + promiseFunctionPrefix
+    },
+    set: {
+      callbackFuncName: "set" + method,
+      promiseFuncName: "set" + method + promiseFunctionPrefix
```
Comment (dxg) [inline lib/LazyLoad.js]: We already know that this object has function names inside, so we can make it shorter:
```js
functionNames.remove.callback
functionNames.remove.promise
```
The context explains that it's a function name.

File: lib/Settings.js [hunk 9]
```diff
@@ -23,7 +23,8 @@ var default_settings = {
     reconnect                 : true,
     pool                      : false,
     debug                     : false
-  }
+  },
+  promiseFunctionPrefix       : 'Async'
```
Comment (dxg) [inline lib/Settings.js]: http://www.differencebetween.com/difference-between-prefix-and-vs-postfix/

**Pull request merged**